### ISSUE TITLE
High PCC but High ATOL/RTOL in Conv2D Output

### ISSUE THREAD

**Issue by mstojkovicTT** [tenstorrent/tt-metal#22314]
### Describe the bug

There are certain Conv2D setups where PCC is high (>0.9), but the absolute tolerance (`atol`) and relative tolerance (`rtol`) are significantly higher than expected. In such cases, the entire model misbehaves as a result.

This issue was initially discussed in [Issue #1935](https://github.com/tenstorrent/tt-forge-fe/issues/1935), where it was determined that the Conv2D operation is the primary contributor to the observed inaccuracies.

Additionally, I checked the predicted classes for ResNet-18 that were provided, and it seems that some of the predicted classes were in the wrong order, meaning the model behaves differently than it should.

### Steps to reproduce the issue

1. Checkout `mstojkovic/conv2d_atol_bug`
2. Download weights: [conv_weights.pt.zip](https://github.com/user-attachments/files/20299978/conv_weights.pt.zip)
3. Run `pytest -svv conv2d_repro.py`

or:

Run the following script:
```python
import torch
import ttnn
import numpy as np
from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import get_atol_rtol_pcc

L1_SMALL_SIZE = 1 << 15
torch.manual_seed(0)

def ttnn_out(input_torch, weights_ttnn, device):
    v4 = input_torch.transpose(1, 2)
    v5 = v4.transpose(2, 3)
    v6 = v5.reshape([1, 1, 196, 256])
    v7 = ttnn.from_torch(v6, device=device, layout=ttnn.TILE_LAYOUT, memory_config=ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.DRAM))
    v8 = ttnn.to_layout(v7, ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED))
    v9 = ttnn.to_device(v8, device, memoDeviceComputeKernelConfigry_config=ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.DRAM))
    
    # Conv2d operation
    v11 = ttnn.conv2d(
        input_tensor=v9,
        weight_tensor=weights_ttnn,
        device=device, 
        out_channels=512, 
        in_channels=256,
        batch_size=1,
        input_height=14, 
        input_width=14,
        kernel_size=(3, 3),
        stride=(2, 2),
        padding=(1, 1),
        dilation=(1, 1),
        groups=1,
        conv_config=ttnn.Conv2dConfig(dtype=ttnn.float32, weights_dtype=ttnn.float32),
        memory_config=ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.DRAM)
    )

    v11 = ttnn.to_torch(v11)
    v12 = v11.reshape([1, 7, 7, 512])
    v13 = v12.transpose(2, 3)
    v14 = v13.transpose(1, 2)
    return v14

def torch_out(input_torch, weights_torch):
    # PyTorch Conv2d operation
    conv = torch.nn.Conv2d(
        in_channels=256, 
        out_channels=512, 
        kernel_size=(3, 3), 
        stride=(2, 2), 
        padding=(1, 1), 
        dilation=(1, 1), 
        groups=1, 
        bias=False
    )
    
    # Load weights
    conv.weight.data = weights_torch
    
    # Forward pass
    output = conv(input_torch)
    
    return output

def create_inputs_for_forward(device):
    # Create input tensor
    shape = (1, 256, 14, 14)
    input_torch = torch.randn(shape, dtype=torch.float32)
    
    # Load weights from file
    weights_torch = torch.load("conv_weights.pt")
    weights_ttnn = ttnn.from_torch(weights_torch, device=None, layout=ttnn.ROW_MAJOR_LAYOUT)
    
    return input_torch, weights_torch, weights_ttnn

def test_conv2d():
    # Set device
    device_id = 0
    device = ttnn.open_device(device_id=device_id, l1_small_size=L1_SMALL_SIZE)
    
    # Get input tensors and weights
    input_torch, weights_torch, weights_ttnn = create_inputs_for_forward(device)
    
    # Compute ttnn output
    tt_out = ttnn_out(input_torch, weights_ttnn, device)
    
    # Compute PyTorch output
    pytorch_out = torch_out(input_torch, weights_torch)
    
    # Calc rtol and atol
    print(f"PyTorch output: {pytorch_out}")
    print(f"TTNN output: {tt_out}")
    _, _, device_pcc, pcc_str = get_atol_rtol_pcc(pytorch_out, tt_out)
    print(f"Output: {pcc_str}")
```

### Expected behavior

current values of `atol` and `rtol` are higher than expected, and they should be significantly lower to ensure numerical accuracy

**Comment by mstojkovicTT** [comment]
cc: @nvukobratTT

**Comment by mstojkovicTT** [comment]
ps. this is a community issue, so it should be prioritized.

**Comment by skrsticTT** [comment]
Hey @mstojkovicTT, a few questions/requests:

- Could you attach a branch for this repro with a test command, so it doesn't become obsolete?

- Can you prioritize this issue (put priority label) for better visibility if it's something you'd like us to focus on? Also, please add the forge label

- Did you try adjusting the parameters in `DeviceComputeKernelConfig` to improve precision (setting `math_fidelity` to `MathFidelity::HiFi4`, `fp32_dest_acc_en` to `true`, `packer_l1_acc` to `true`)?

- How did you determine that this conv2d is causing the problem? I know a good PCC doesn't always mean everything is fine, but did you try using the reference implementation to compare the results? If so, could you share the results of the reference model vs. the TTNN model?

**Comment by mstojkovicTT** [comment]
> Could you attach a branch for this repro with a test command, so it doesn't become obsolete?

Added to desc under repro part

> Can you prioritize this issue (put priority label) for better visibility if it's something you'd like us to focus on? Also, please add the forge label

Added forge label, but didnt want to put p-priority myself, this is p0 on Forge-FE side

> How did you determine that this conv2d is causing the problem? I know a good PCC doesn't always mean everything is fine, but did you try using the reference implementation to compare the results? If so, could you share the results of the reference model vs. the TTNN model?

Here is the label prediction between Compiler model and Pytorch model:
```
Compiler Top-5 Predictions:
  Class 671: 0.6997
  Class 870: 0.1314
  Class 880: 0.1281
  Class 444: 0.0215
  Class 665: 0.0153
PyTorch Top-5 Predictions:
  Class 671: 0.6359
  Class 880: 0.1640
  Class 870: 0.1596
  Class 444: 0.0221
  Class 665: 0.0133
```

As you can see, `Class 880` and `Class 870` even changed place.

**Comment by skrsticTT** [comment]
I see that the compiler produces different results compared to PyTorch, but what indicates that this specific conv2d is the problem? What happens in the compiler's model if you replace this ttnn.conv2d with the one from PyTorch?

**Comment by mstojkovicTT** [comment]
I haven't replaced the `ttnn.conv2d` ops in the model directly. I just extracted specific ops that are generated in the `ResNet18` graph, ran them independently, and observed a spike in `conv2d` `rtol` and `atol`

**Comment by pavlejosipovic** [comment]
Running your test case as is, I"m getting following output
```
Output: Max ATOL Delta: 0.1950528621673584, Max RTOL Delta: 200.57395935058594, PCC: 0.9995897034363761
end
```
After setting up "the best" compute config like this

```
compute_config = ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.HiFi4,
        fp32_dest_acc_en=True,
        packer_l1_acc=True,
    )

  # Conv2d operation
  v11 = ttnn.conv2d(
      input_tensor=v9,
      weight_tensor=weights_ttnn,
      device=device, 
      out_channels=512, 
      in_channels=256,
      batch_size=1,
      input_height=14, 
      input_width=14,
      kernel_size=(3, 3),
      stride=(2, 2),
      padding=(1, 1),
      dilation=(1, 1),
      groups=1,
      conv_config=ttnn.Conv2dConfig(dtype=ttnn.float32, weights_dtype=ttnn.float32),
      memory_config=ttnn.MemoryConfig(ttnn.TensorMemoryLayout.INTERLEAVED, ttnn.BufferType.DRAM),
      compute_config=compute_config,
  )
```
I'm getting following output
```
Output: Max ATOL Delta: 0.0061604976654052734, Max RTOL Delta: 9.703466415405273, PCC: 0.9999997999670269
end
```

There is not bug here from conv op perspective.
